[FLINK-17733][FLINK-16448][hive][doc] Adjust Hive doc & Add documentation for real-time hive (#12537)
apache/flink: docs/dev/table/hive/hive_streaming.md

and runs a batch query to read that data back out. 

{% highlight sql %}

SET table.sql-dialect=hive;
CREATE TABLE hive_table (
  user_id STRING,
  order_amount DOUBLE
) PARTITIONED BY (dt STRING, hour STRING) STORED AS parquet TBLPROPERTIES (
  'partition.time-extractor.timestamp-pattern'='$dt $hour:00:00',
  'sink.partition-commit.trigger'='partition-time',
  'sink.partition-commit.delay'='1 h',
  'sink.partition-commit.policy.kind'='metastore,success-file'
);

SET table.sql-dialect=default;
CREATE TABLE kafka_table (
  user_id STRING,
  order_amount DOUBLE,
  log_ts TIMESTAMP(3),
  WATERMARK FOR log_ts AS log_ts - INTERVAL '5' SECOND
) WITH (...);

-- streaming sql, insert into hive table
INSERT INTO TABLE hive_table SELECT user_id, order_amount, DATE_FORMAT(log_ts, 'yyyy-MM-dd'), DATE_FORMAT(log_ts, 'HH') FROM kafka_table;

-- batch sql, select with partition pruning
SELECT * FROM hive_table WHERE dt='2020-05-20' and hour='12';

{% endhighlight %}

## Streaming Reading

To improve the real-time performance of hive reading, Flink support real-time Hive table stream read:

> - Partition table, monitor the generation of partition, and read the new partition incrementally.

Reviewer: We need to mention that new data added to an existing partition won't be consumed

Author: I will explain more in NOTE.

- Non-partition table, monitor the generation of new files in the folder, and read new files incrementally.

You can even use the 10 minute level partition strategy, and use Flink's Hive streaming reading and
Hive streaming writing to greatly improve the real-time performance of Hive data warehouse to quasi
real-time minute level.

<table class="table table-bordered">
  <thead>
    <tr>
        <th class="text-left" style="width: 20%">Key</th>
        <th class="text-left" style="width: 15%">Default</th>
        <th class="text-left" style="width: 10%">Type</th>
        <th class="text-left" style="width: 55%">Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
        <td><h5>streaming-source.enable</h5></td>
        <td style="word-wrap: break-word;">false</td>
        <td>Boolean</td>

>         <td>Enable streaming source or not. NOTES: For non-partition table, please make sure that each file should be put atomically into the target directory, otherwise the reader may get incomplete data.</td>

Reviewer: For partitioned table, I think each partition should also be added atomically right?

Author: Yes, will be modified to:
`Enable streaming source or not.NOTES: Please make sure that each partition/file should be written atomically, otherwise the reader may get incomplete data.`

    </tr>
    <tr>
        <td><h5>streaming-source.monitor-interval</h5></td>
        <td style="word-wrap: break-word;">1 m</td>
        <td>Duration</td>
        <td>Time interval for consecutively monitoring partition/file.</td>
    </tr>
    <tr>
        <td><h5>streaming-source.consume-order</h5></td>
        <td style="word-wrap: break-word;">create-time</td>
        <td>String</td>

>         <td>The consume order of streaming source, support create-time and partition-time. create-time compare partition/file creation time, this is not the partition create time in Hive metaStore, but the folder/file create time in filesystem; partition-time compare time represented by partition name. For non-partition table, this value should always be 'create-time'.</td>

Reviewer: I think we should warn the users that the `create-time` here is actually the modification time in Hadoop FS. So if the partition folder somehow gets updated, e.g. changing ACL attributes, it can affect how the data is consumed.
And perhaps we should make `partition-time` as the default and recommended value for this property?

Author: Default `partition-time` +1

Reviewer: I see. If we change default value to partition-time and user is using a non-partitioned table then the query will fail, right? If that's the case, then maybe we can go with create-time as default, but with proper warnings.

    </tr>
    <tr>
        <td><h5>streaming-source.consume-start-offset</h5></td>
        <td style="word-wrap: break-word;">1970-00-00</td>
        <td>String</td>

>         <td>Start offset for streaming consuming. How to parse and compare offsets depends on your order. For create-time and partition-time, should be a timestamp string. For partition-time, will use partition time extractor to extract time from partition.</td>

Reviewer: Mention the supported timestamp format is `yyyy-[m]m-[d]d [hh:mm:ss]`?

    </tr>
  </tbody>
</table>

Note:

> - Monitor strategy is scan all directories/files in location path now. If there is too more partitions, will have performance problems.

Reviewer: ```suggestion
- Monitor strategy is to scan all directories/files in location path now. If there are too many partitions, there will be performance problems.
```

- Streaming reading requires that the file or partition is atomic in the view of hive metastore and does not support append writing.
- Streaming reading not support watermark grammar in Flink DDL. So it can not be used for window operators.

The below shows how to read Hive table incrementally. 

{% highlight sql %}

> SELECT * FROM hive_table /*+ OPTIONS('streaming-source.enable'=’true’, 'streaming-source.consume-start-offset'='2020-05-20') */;

Author: ’true’ -> `'`
